src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,244 @@
+  {D : A → B} (Dm : ∀ {a b}, D a + D b ≤ D (a + b))
+  (f g : add_monoid_algebra R A) :
+  f.support.inf D + g.support.inf D ≤ (f * g).support.inf D :=
+order_dual.of_dual_le_of_dual.mpr $
+  sup_support_mul_le (λ a b, order_dual.of_dual_le_of_dual.mp Dm) f g
+
+end add_only
+
+lemma sup_support_list_prod_le [add_monoid A] [add_monoid B] [semilattice_sup B] [order_bot B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b) :
+  ∀ l : list (add_monoid_algebra R A),
+    l.prod.support.sup D ≤ (l.map (λ f : add_monoid_algebra R A, f.support.sup D)).sum
+| [] := begin
+    rw [list.map_nil, finset.sup_le_iff, list.prod_nil, list.sum_nil],
+    exact λ a ha, by rwa [finset.mem_singleton.mp (finsupp.support_single_subset ha)]
+  end
+| (f::fs) := begin
+    rw [list.prod_cons, list.map_cons, list.sum_cons],
+    exact (sup_support_mul_le Dm _ _).trans (add_le_add_left (sup_support_list_prod_le _) _)
+  end
+
+lemma list_prod_le_sup_support_sum [add_monoid A] [add_monoid B] [semilattice_inf B] [order_top B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)] {D : A → B}
+  (D0 : 0 ≤ D 0) (Dm : ∀ a b, D a + D b ≤ D (a + b)) (l : list (add_monoid_algebra R A)) :
+  (l.map (λ f : add_monoid_algebra R A, f.support.inf D)).sum ≤ l.prod.support.inf D :=
+order_dual.of_dual_le_of_dual.mpr $ sup_support_list_prod_le (order_dual.of_dual_le_of_dual.mp D0)
+  (λ a b, order_dual.of_dual_le_of_dual.mp (Dm _ _)) l
+
+lemma sup_support_pow_le [add_monoid A] [add_monoid B] [semilattice_sup B] [order_bot B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b) (n : ℕ)
+  (f : add_monoid_algebra R A) :
+  (f ^ n).support.sup D ≤ n • (f.support.sup D) :=
+begin
+  rw [← list.prod_repeat, ←list.sum_repeat],
+  refine (sup_support_list_prod_le D0 Dm _).trans_eq _,
+  rw list.map_repeat,
+end
+
+lemma nsmul_inf_support_le_pow [add_monoid A] [add_monoid B] [semilattice_inf B] [order_top B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+  {D : A → B} (D0 : 0 ≤ D 0) (Dm : ∀ a b, D a + D b ≤ D (a + b)) (n : ℕ)
+  (f : add_monoid_algebra R A) :
+  n • (f.support.inf D) ≤ (f ^ n).support.inf D :=
+order_dual.of_dual_le_of_dual.mpr $ sup_support_pow_le (order_dual.of_dual_le_of_dual.mp D0)
+  (λ a b, order_dual.of_dual_le_of_dual.mp (Dm _ _)) n f
+
+end semiring
+
+variables [comm_semiring R] [add_comm_monoid A] [add_comm_monoid B]
+
+lemma sup_support_multiset_prod_le [semilattice_sup B] [order_bot B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b)
+  (F : multiset (add_monoid_algebra R A)) :
+  F.prod.support.sup D ≤ (F.map (λ f : add_monoid_algebra R A, f.support.sup D)).sum :=
```

Comment:
`F` is an odd name for a multiset too :). Maybe `m`?

Reply:
Sure, I'll change it to `m`!